src/serde/ser.rs

```diff
@@ -1,12 +1,75 @@
+use std::fmt::Arguments;
 use std::io;
 use std::io::Cursor;
+use std::io::ErrorKind;
+use std::io::IoSlice;
+use std::io::Write;
 
 use super::write_atom::write_atom;
 use crate::allocator::{NodePtr, SExp};
 use crate::node::Node;
 
 const CONS_BOX_MARKER: u8 = 0xff;
 
+pub struct LimitedWriter<W: io::Write> {
+    inner: W,
+    limit: i64,
+}
+
+impl<W: io::Write> LimitedWriter<W> {
+    pub fn new(w: W, limit: i64) -> LimitedWriter<W> {
+        LimitedWriter { inner: w, limit }
+    }
+
+    pub fn into_inner(self) -> W {
+        self.inner
+    }
+}
+
+impl<W: io::Write> Write for LimitedWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if (self.limit - buf.len() as i64) < 0 {
+            return Err(ErrorKind::OutOfMemory.into());
```

Comment:
Consider

```suggestion
            return Err(ErrorKind:: StorageFull.into());
```

Reply:
I'm not sure about this. Today there's no use case for receiving an error *and* use the partial output for something. I think making this change (and adding this complexity) is definitely fine to do, as soon as we have a use case for it. (I subscribe to the "you ain't gonna need it" principle)

Reply:
`StorageFull` is not part of the stable API yet (as far as I understand). It requires building with the `io_error_more` feature. This makes me hesitate to use it for now.

Reply:
Ah okay. I hadn't tried it.

I was looking around for more options, when I saw `WriteZero` and realized, it's probably enough for `write` to simply return `0` each time and let `write_fully` return the `WriteZero` error. I'll try it locally.

Reply:
Yep, that works! In conjunction with the `min` code path, we get everything we need.